cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/RunRecordCounter.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.services;
+
+import io.cdap.cdap.app.runtime.ProgramRuntimeService;
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.ProgramType;
+import io.cdap.cdap.proto.id.ProgramRunId;
+import org.eclipse.jetty.util.ConcurrentHashSet;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Helper class to maintain and return total number of launching and running run-records.
+ */
+public class RunRecordCounter {
+  /**
+   * Contains ProgramRunIds of runs that have been accepted, but have not been added to metadata store plus
+   * all run records with {@link ProgramRunStatus#PENDING} status.
+   */
+  private final Set<ProgramRunId> inflightAndPendings;
+  private final ProgramRuntimeService runtimeService;
+
+  public RunRecordCounter(ProgramRuntimeService runtimeService) {
+    this.runtimeService = runtimeService;
+
+    inflightAndPendings = new ConcurrentHashSet<>();
+  }
+
+  /**
+   * Add a new inflight launch request and return total number of launching and running program runs.
+   *
+   * @param programRunId run id associated with the launch request
+   * @return total number of launching and running program runs.
+   */
+  public Count addRequestAndGetCount(ProgramRunId programRunId) {
+    int inflightAndPendingCount;
+    synchronized (this) {
```

Comment:
Do you really need synchronized here? Are there difference for some other add happened before right synchronized and between `add` and `size`?

Reply:
Let's assume we have 1 more slot for accepting a new launch request and two requests come in concurrently. We clearly would like to accept one and reject the other one. 

Basically without this synchronized block, concurrent requests will be accepted by the system regardless of set max threshold.

Reply:
Not really. It's actually visa versa :)
The only thing thing synchronized block guards is the ability to do something between `add` and `size`.
So next is possible:
Max size is N+1
1. Thread 1 add
2. Thread 2 add
3. Thread 2 size = N+2, reject
4. Thread 1 size = N+2, reject
While it's not good, I am wondering if it's worth `synchronized`. `synchronized` is actually a very heavy thing under contention

Reply:
Discussed this offline with Vitalii. Agreed to include synchronization given its benefits and despite added overhead.